demo.ipynb

```diff
@@ -10,6 +10,94 @@
         "<a href=\"https://colab.research.google.com/github/geraschenko/gmrfs/blob/main/demo.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
       ]
     },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# Fitting a latent 2d Gaussian Markov Random Field\n",
+        "\n",
+        "We consider a latent 2d GMRF known as the [Intrinsic Conditional Autoregressive (ICAR) model](https://mc-stan.org/users/documentation/case-studies/icar_stan.html) which has the form\n",
+        "\\begin{align}\n",
+        "p(x|c,\\theta) &= N(x|\\mu,Q^{-1}) \\\\\n",
+        "\\mu &= c^T \\beta \\\\\n",
+        "Q &= \\tau(D - W)\n",
+        "\\end{align}\n",
+        "where $x$ is a set of $n$ latent scalars,\n",
+        "$c$ are covariates,\n",
+        "$\\theta=(\\tau,\\beta)$ are the parameters,\n",
+        "$D$ is the $n \\times n$ diagonal matrix where entries $(i,i)$\n",
+        "are the number of neighbors of region $i$\n",
+        "and the off-diagonal entries are 0,\n",
+        "and\n",
+        "$W$ is the $n \\times n$ adjacency matrix where entries $(i,i)$\n",
+        "are zero and the off-diagonal elements are 1 if regions $i$ and $j$\n",
+        "are neighbors and 0 otherwise.\n",
+        "Note that $Q$ is sparse, and this prior is singular (improper).\n",
```

Comment:
It may be worth pointing out that in the implementation, we add diagonal regularization to make it non-singular (see https://github.com/geraschenko/gmrfs/blob/main/gmrfs/gmrf.py#L116).

Reply:
Addressed in https://github.com/geraschenko/gmrfs/pull/2/commits/cdc292c0fd70e73ce3d56a5d2b26bd4402382e31

Reply:
Yes, this is worth mentioning.